### PR TITLE
apps wc: Propagate PSA ns labels for HNC namespaces

### LINKED ISSUE #1530: [2] Automatically assign restricted PSA to customer namespaces
**Is your feature request related to a problem? Please describe.**
Part of PSP to PSA.

We need to have a way of automatically assigning the restricted PSA policy to all namespaces that customers create.

**Describe the solution you'd like**
I see two options
1. Gatekeeper mutating webhook for adding the labels for all non-labeled namesapces.
2. Configure managed labels via HNC
    - [use-managed-labels](https://github.com/kubernetes-sigs/hierarchical-namespaces/blob/master/docs/user-guide/how-to.md#use-managed-labels)

I prefer option 2.

Be careful and determine if there could be any breaking side-effects by using managed labels, also if it might break any existing setup.

** Additional Context **
It might be worth to keep in mind to see what we can do if we ever end up in the situation where another but the restricted PSA needs to be assigned to a customer namespace.

**Definition of done:**
- [ ] Customer created namespaces are automatically assigned the restricted PSA labels.

### PR DESCRIPTION
**Pull request by robinAwallace**
**What this PR does / why we need it**:

1. Propagates PSA labels for user hnc namespaces
2. Possible to override usernamespace PSA level (Will not be a managed HNC namespace)
3. Moved `read-hierarchyconfigurations-role` into hnc user-rbac
4. Removed hnc admin-rbac from admin (attached to user admins)

**Which issue this PR fixes**: fixes #1530 

**Public facing documentation PR** *(if applicable)*

**Special notes for reviewer**:

**Add a screenshot or an example to illustrate the proposed solution:**

**Checklist:**

- [x] Added relevant notes to [WIP-CHANGELOG.md](https://github.com/elastisys/compliantkubernetes-apps/blob/main/WIP-CHANGELOG.md)
- [x] Proper commit message prefix on all commits
- [ ] Updated the [public facing documentation](https://github.com/elastisys/compliantkubernetes)
- Is this changeset backwards compatible for existing clusters? Applying:
    - [ ] is completely transparent, will not impact the workload in any way.
    - [x] requires running a migration script.
    - [ ] will create noticeable cluster degradation.
          E.g. logs or metrics are not being collected or Kubernetes API server
          will not be responding while upgrading.
    - [ ] requires draining and/or replacing nodes.
    - [ ] will change any APIs.
          E.g. removes or changes any CK8S config options or Kubernetes APIs.
    - [ ] will break the cluster.
          I.e. full cluster migration is required.
- Chart checklist (pick exactly one):
    - [x] I upgraded no Chart.
    - [ ] I upgraded a Chart and determined that no migration steps are needed.
    - [ ] I upgraded a Chart and added [migration steps](https://github.com/elastisys/compliantkubernetes-apps/blob/main/migration).

**Pipeline config** *(if applicable)*
If you change some config options (e.g. add/rename variable or change the default value) you may need to update the config used by the pipeline in `pipeline/config`.

### CODE REVIEW

File: WIP-CHANGELOG.md [hunk 1]
```diff
@@ -22,6 +22,10 @@
 - Add Gatekeeper PSPs for HNC.
 - Add Gatekeeper PSPs for falco.
 - Add cache-image workflow
+- Add PSA labels propagate for user hnc owned namespaces
+- Add possible to override usernamespace `restricted` PSA level
```
Comment (OlleLarsson) [inline WIP-CHANGELOG.md:26]:
```suggestion
- Add option to override usernamespace `restricted` PSA level
```

File: helmfile/values/user-rbac.yaml.gotmpl [hunk 2]
```diff
@@ -1,4 +1,14 @@
-namespaces: {{ toYaml .Values.user.namespaces | nindent 2 }}
+namespaces:
+{{- range $namespace := .Values.user.namespaces }}
+  - name: {{ $namespace -}}
+    {{ if (hasKey $.Values.user.constraints $namespace ) }}
+    {{- $psa := (index $.Values.user.constraints $namespace) }}
+    psa: {{ $psa.psaLevel | quote }}
+    {{ else }}
+    psa: "restricted"
```
Comment (OlleLarsson) [inline helmfile/values/user-rbac.yaml.gotmpl:8]: Are quotes needed?
```suggestion
    psa: {{ $psa.psaLevel }}
    {{ else }}
    psa: restricted
```

File: WIP-CHANGELOG.md [hunk 3]
```diff
@@ -71,9 +75,12 @@
...
 - Install Gatekeeper PSPs per stack to improve targeting and modularity
+- Moved HNC `read-hierarchyconfigurations-role` into HNC user-rbac.
 
 ### Removed
 
 - Kubernetes PSP for ingress-nginx and monitoring namespace.
 - Remove Kubernetes PSPs for Fluentd and OpenSearch
 - Remove Kubernetes PSPs for falco
+- Remove HNC admin-rbac from admin (attached to user admins)
+  - User admins will not only have the HNC user-rbac instead.
```
Comment (OlleLarsson) [inline WIP-CHANGELOG.md:86]: ?
```suggestion
  - User admins will now only have the HNC user-rbac instead.
```

File: helmfile/charts/hnc/controller/templates/user-rbac.yaml [hunk 4]
```diff
@@ -7,6 +7,18 @@ metadata:
...
+  resources:
+  - hierarchyconfigurations
+  verbs:
+  - get
+  - list
+  - watch
+{{- if $.Values.rbac.user.allowConfigEdit }}
+  - patch
+  - update
+{{- end }}
```
Comment (aarnq) [inline helmfile/charts/hnc/controller/templates/user-rbac.yaml:21]: Shouldn't this be per namespace? (Or at least the edit part?)

**Comment by aarnq** [comment]
We should tie up all dependencies we now have on gatekeeper. Or make gatekeeper non optional. (Though that should be done in a separate PR.)

File: helmfile/charts/hnc/controller/templates/user-rbac.yaml (hunk 4 above)
Comment (robinAwallace) [reply helmfile/charts/hnc/controller/templates/user-rbac.yaml:21]: Yes the edit/patch should be on a namespace level. I guess we said it is fine for them to list all namespaces and that includes hnc namepsaces

Comment (robinAwallace) [reply helmfile/charts/hnc/controller/templates/user-rbac.yaml:21]: Yes I think i will remove it

File: migration/v0.30/apply/30-remove-starboard.sh [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+      helm_uninstall ${CLUSTER} monitoring starboard-operator
+      helm_uninstall ${CLUSTER} monitoring starboard-psp-rbac
+      log_info "Delete starboard crds"
+
+      STAR_CRDS=$(
+        kubectl_do ${CLUSTER} api-resources --api-group=aquasecurity.github.io -o name
+      )
+      if [ -n "$STAR_CRDS" ]; then
+          # shellcheck disable=SC2086
+          kubectl_do ${CLUSTER} delete crds --ignore-not-found=true $STAR_CRDS
```
Comment (robinAwallace) [inline migration/v0.30/apply/30-remove-starboard.sh:22]: @aarnq I just realized that this is not super good, as running this again will remove the new CRDs aswell :disappointed:
I guess i must check the actual CRD version and if that is the old version delete it...

Comment (robinAwallace) [reply migration/v0.30/apply/30-remove-starboard.sh]: Fixed

File: migration/v0.30/apply/30-remove-starboard.sh [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+      if [ -n "$STAR_CRDS" ]; then
+          # shellcheck disable=SC2086
+          for STAR_CRD in $STAR_CRDS;
+          do
+              echo $STAR_CRD
+          done
+          kubectl_do ${CLUSTER} delete --ignore-not-found=true $STAR_CRDS
+      fi
+    done
+
```
Comment (aarnq) [inline migration/v0.30/apply/30-remove-starboard.sh:30]:
```suggestion
    for CLUSTER in sc wc; do
      log_info "Uninstall starboard"
      helm_uninstall ${CLUSTER} monitoring starboard-operator
      helm_uninstall ${CLUSTER} monitoring starboard-psp-rbac
      log_info "Delete starboard crds"

      STAR_CRDS=$(kubectl_do ${CLUSTER} get crds -l app.kubernetes.io/managed-by=starboard -oname)

      if [ -n "$STAR_CRDS" ]; then
          # shellcheck disable=SC2086
          for STAR_CRD in $STAR_CRDS; do
              echo $STAR_CRD
          done
          kubectl_do ${CLUSTER} delete --ignore-not-found=true $STAR_CRDS
      fi
    done

```

**Review by aarnq — APPROVED** [review]

**Review by viktor-f — APPROVED** [review]

**Pull request merged**